homeassistant/components/deluge/__init__.py

```diff
@@ -1 +1,356 @@
+        if "Username does not exist" in str(error):
+            raise UserNameError
+        if "Password does not match" in str(error):
+            raise PasswordError
+        raise UnknownError
+
+
+class DelugeClient:
+    """Deluge Client Object."""
+
+    def __init__(self, hass, config_entry):
+        """Initialize the Deluge client."""
+        self.hass = hass
+        self.config_entry = config_entry
+        self._deluge_data = None
+        self.unsub_timer = None
+
+    @property
+    def api(self):
+        """Return the _deluge_data object."""
+        return self._deluge_data
+
+    async def async_setup(self):
+        """Set up the Deluge client."""
+
+        try:
+            deluge_api = await get_api(self.hass, self.config_entry.data)
+        except CannotConnect:
+            raise ConfigEntryNotReady
+        except (UserNameError, PasswordError, UnknownError):
+            return False
+
+        self._deluge_data = DelugeData(self.hass, self.config_entry, deluge_api)
+
+        await self.hass.async_add_executor_job(self._deluge_data.init_torrent_list)
+        await self.hass.async_add_executor_job(self._deluge_data.update)
+        self.add_options()
+        self.set_scan_interval(self.config_entry.options[CONF_SCAN_INTERVAL])
+
+        for component in PLATFORMS:
+            self.hass.async_create_task(
+                self.hass.config_entries.async_forward_entry_setup(
+                    self.config_entry, component
+                )
+            )
+
+        def add_torrent(service):
+            """Add new torrent to download."""
+            torrent = service.data[ATTR_TORRENT]
+            deluge_client = None
+            for entry in self.hass.config_entries.async_entries(DOMAIN):
+                if entry.data[CONF_HOST] == service.data[CONF_HOST]:
+                    deluge_client = self.hass.data[DOMAIN][entry.entry_id]
+                    break
+            if deluge_client is None:
+                _LOGGER.error("Deluge host is not found")
+                return
+            if torrent.startswith(("http", "ftp:")):
+                deluge_api.core.add_torrent_url(torrent, {})
+            elif torrent.startswith(("magnet:")):
+                deluge_api.core.add_torrent_magnet(torrent, {})
+            else:
+                _LOGGER.warning("Could not add torrent: unsupported type")
+
+        self.hass.services.async_register(
+            DOMAIN, SERVICE_ADD_TORRENT, add_torrent, schema=SERVICE_ADD_TORRENT_SCHEMA
+        )
+
+        self.config_entry.add_update_listener(self.async_options_updated)
+
+        return True
+
+    def add_options(self):
+        """Add options for entry."""
+        if not self.config_entry.options:
+            scan_interval = self.config_entry.data.pop(
+                CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
+            )
+            options = {CONF_SCAN_INTERVAL: scan_interval}
+
+            self.hass.config_entries.async_update_entry(
```

Comment:
async method from sync context.

Reply:
We add `async` to the function name for a reason 😉

Reply:
I see that but I have noticed that same has been done in other integration. Do you recommend to make the add_options method async?

Reply:
We're calling `add_options` from async context. It's just the name of the method that is a bit misleading.